### PR TITLE
Fix memory leak in 'NavigationServer3D' involving static obstacles

### PR DESCRIPTION
**Pull request by Scony**
This PR was about adding further core tests for `NavigationServer3D`, but it exposed quite a serious memory leak, so I've extended implementation to fix that.

CC @smix8

### CODE REVIEW

**Comment by Scony** [comment]
@akien-mga this PR is now a bugfix with tests - if you like I can extract tests to separate PR.

**Review by smix8 — APPROVED** [review]

File: tests/servers/test_navigation_server_3d.h [hunk 1]
```diff
@@ -429,6 +429,93 @@ TEST_SUITE("[Navigation]") {
...
+		navigation_server->obstacle_set_map(obstacle_1, map);
+		navigation_server->obstacle_set_avoidance_enabled(obstacle_1, true);
+		navigation_server->obstacle_set_position(obstacle_1, Vector3(2.5, 0, 0.5));
+		navigation_server->obstacle_set_radius(obstacle_1, 1);
+
+		CHECK_EQ(agent_1_avoidance_callback_mock.function1_calls, 0);
+		navigation_server->process(0.0); // Give server some cycles to commit.
+		CHECK_EQ(agent_1_avoidance_callback_mock.function1_calls, 1);
+		Vector3 agent_1_safe_velocity = agent_1_avoidance_callback_mock.function1_latest_arg0;
+		CHECK_MESSAGE(agent_1_safe_velocity.x > 0, "agent 1 should move a bit along desired velocity (+X)");
```
[the remark this answers is not in the document]
Comment (Scony) [reply tests/servers/test_navigation_server_3d.h]: Done

**Review by AThousandShips — APPROVED** [review]

**Comment by akien-mga** [comment]
Thanks!

**Comment by YuriSizov** [comment]
Cherry-picked for 4.2.1.

**Pull request merged**